### PR TITLE
Database set-up

### PR DESCRIPTION
**Pull request by gdetrez**
This is a partial implementation of a persistence layer using PostgreSQL. The idea is to discuss the approach before I set-up all the tables and change the corresponding code.
What's included:
- install dependencies (sqlalchemy, psycopg...)
- set-up package structure: I created a new package instead of using `scanomatic.io` to group the database related modules together.
- setup flask to connect to the database
- add postgres to docker-compose
- set-up a migration framework
- update the docker image to execute migrations when starting
- add a table for the scanners
- add a gateway class to save/fetch scanners from the database

Some notes:
- Running the migration in the docker entrypoint makes things simple right now but it may not be the most efficient thing.  And the container will not start if the database is not reachable.
- I choose to not use the sqlalchemy ORM. This works well with the current models and will hopefully be more flexible when we will want to store arrays in the database.
- on the python side my idea is to split `ScanningStore`, loosely following the different APIs (scanner, scan, scanjob...)

### CODE REVIEW

File: tests/unit/ui_server/test_database.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+from flask import Flask
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply tests/unit/ui_server/test_database.py]: fixed

File: scanomatic/data/migrations/versions/6736f58587af_create_scanners_table.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+"""Create scanners table
+
+Revision ID: 6736f58587af
+Revises:
+Create Date: 2018-02-21 10:12:27.068506
+
+"""
+from alembic import op
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply scanomatic/data/migrations/versions/6736f58587af_create_scanners_table.py]: wontfix: this is generated code

File: scanomatic/data/scannerstore.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import sqlalchemy as sa
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply scanomatic/data/scannerstore.py]: fixed

File: scanomatic/data/tables.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from sqlalchemy import Table, MetaData, Column, Text
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply scanomatic/data/tables.py]: :heavy_check_mark:

File: scanomatic/ui_server/database.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+from collections import namedtuple
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply scanomatic/ui_server/database.py]: :heavy_check_mark:

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import os
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply tests/integration/conftest.py]: :heavy_check_mark:

File: tests/integration/data/test_scannerstore.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
+import pytest
```
[the remark this answers is not in the document]
Comment (gdetrez) [reply tests/integration/data/test_scannerstore.py]: :heavy_check_mark:

File: scanomatic/ui_server/database.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        if dbstate is None:
+            return
+        if exception is None:
+            dbstate.transaction.commit()
+        else:
+            dbstate.transaction.rollback()
+
+
+def connect():
+    global x
```
Comment (local-minimum) [inline scanomatic/ui_server/database.py]: remove

**Comment by gdetrez** [comment]
@local-minimum @dindinoh : I made a slight change to the way the database is configured: using multiple environment variable (`PGHOST`, `PGUSER`...) instead of one and allowing us to store the database password in `/etc/scanomatic/pgpass`.

File: docker-compose.yml [hunk 9]
```diff
@@ -11,12 +11,13 @@ services:
       - somprojects:/somprojects
       - /dev/bus/usb:/dev/bus/usb
     environment:
-      - DATABASE_URL=postgresql://postgres:password@db/scanomatic
+      - PGHOST=db
+      - PGUSER=postgres
+      - PGDATABASE=scanomatic
```
Comment (local-minimum) [inline docker-compose.yml]: Detail: here it is `PGDATABASE` but further down `POSTGRES_DB`...

**Pull request merged**